src/data/sign.lean

```diff
@@ -0,0 +1,196 @@
+  exists_pair_ne  := ⟨0, 1,   by rintro ⟨⟩⟩,
+  inv_zero        := rfl }
+
+instance : linear_order sign_type :=
+{ le           := (≤),
+  le_refl      := λ a, by cases a; constructor,
+  le_total     := λ a b, by casesm* _; dec_trivial,
+  le_antisymm  := λ a b ha hb, by casesm* _; refl,
+  le_trans     := λ a b c hab hbc, by casesm* _; constructor,
+  decidable_le := le.decidable_rel }
+
+instance : has_distrib_neg sign_type :=
+{ neg_neg := λ x, by cases x; refl,
+  neg_mul := λ x y, by casesm* _; refl,
+  mul_neg := λ x y, by casesm* _; refl,
+..sign_type.has_neg }
+
+/-- `sign_type` is equivalent to `fin 3`. -/
+def fin3_equiv : sign_type ≃* fin 3 :=
+{ to_fun :=  λ a, a.rec_on 0 (-1) 1,
+  inv_fun := λ a, match a with
+    | ⟨0, h⟩   := 0
+    | ⟨1, h⟩   := 1
+    | ⟨2, h⟩   := -1
+    | ⟨n+3, h⟩ := (h.not_le le_add_self).elim
+  end,
+  left_inv :=  λ a, by cases a; refl,
+  right_inv := λ a, match a with
+    | ⟨0, h⟩   := rfl
+    | ⟨1, h⟩   := rfl
+    | ⟨2, h⟩   := rfl
+    | ⟨n+3, h⟩ := (h.not_le le_add_self).elim
+  end,
+  map_mul' := λ x y, by casesm* _; refl }
+
+/-- Turn a `sign_type` into zero, one, or minus one. This is a coercion instance, but note it is
+only a `has_coe_t` instance: see note [use has_coe_t]. -/
+def cast {α} [has_zero α] [has_one α] [has_neg α] : sign_type → α
+| zero :=  0
+| pos  :=  1
+| neg  := -1
+
+@[simp] lemma cast_zero    {α} [has_zero α] [has_one α] [has_neg α] : cast 0 = (0 : α) := rfl
+@[simp] lemma cast_one     {α} [has_zero α] [has_one α] [has_neg α] : cast 1 = (1 : α) := rfl
+@[simp] lemma cast_neg_one {α} [has_zero α] [has_one α] [has_neg α] : cast (-1) = (-1 : α) := rfl
+
+instance {α} [has_zero α] [has_one α] [has_neg α] : has_coe_t sign_type α := ⟨cast⟩
+
+/-- `sign_type.cast` as a `mul_with_zero_hom`. -/
+@[simps] def cast_hom {α} [mul_zero_one_class α] [has_distrib_neg α] : sign_type →*₀ α :=
+{ to_fun    := cast,
+  map_zero' := rfl,
+  map_one'  := rfl,
+  map_mul'  := λ x y, by cases x; cases y; simp }
+
+end sign_type
+
+variables {α : Type*}
+
+open sign_type
+
+section preorder
+
+variables [has_zero α] [preorder α] [decidable_rel ((<) : α → α → Prop)] {a : α}
+
+/-- The sign of an element is 1 if it's positive, -1 if negative, 0 otherwise. -/
+def sign : α →o sign_type :=
+⟨λ a, if 0 < a then 1 else if a < 0 then -1 else 0, λ a b h, begin
+  dsimp,
+  split_ifs; try {constructor},
+  { cases lt_irrefl 0 (h_1.trans $ h.trans_lt h_3) },
+  { cases h_2 (h_1.trans_le h) },
+  { cases h_2 (h.trans_lt h_4) }
+  end⟩
+
+lemma sign_apply : sign a = ite (0 < a) 1 (ite (a < 0) (-1) 0) := rfl
+
+@[simp] lemma sign_zero : sign (0 : α) = 0 := by simp [sign_apply]
+@[simp] lemma sign_pos (ha : 0 < a) :  sign a = 1 := by rwa [sign_apply, if_pos]
+@[simp] lemma sign_neg (ha : a < 0) : sign a = -1 := by rwa [sign_apply, if_neg $ asymm ha, if_pos]
+
+end preorder
+
+section linear_order
+
+variables [has_zero α] [linear_order α] {a : α}
+
+lemma sign_ne_zero (h : a ≠ 0) : sign a ≠ 0 :=
```

Comment:
A nicer lemma for rewriting would be something like `sign_eq_zero_iff : sign a = 0 ↔ a = 0`, and then this should be an easy consequence of the iff.